File: backend/utils/security.py
import hashlib
from datetime import datetime, timedelta, timezone
from typing import Optional
from fastapi import HTTPException, Security, status
from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer
from jose import JWTError, jwt
from config import settings

security = HTTPBearer()

def hash_password(password: str) -> str:
    """Hash password using SHA-256 for simple portable execution without external C bindings issues."""
    return hashlib.sha256((password + settings.SECRET_KEY).encode('utf-8')).hexdigest()

def verify_password(plain_password: str, hashed_password: str) -> bool:
    return hash_password(plain_password) == hashed_password

def create_access_token(data: dict, expires_delta: Optional[timedelta] = None) -> str:
    to_encode = data.copy()
    if expires_delta:
        expire = datetime.now(timezone.utc) + expires_delta
    else:
        expire = datetime.now(timezone.utc) + timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
    to_encode.update({"exp": expire})
    encoded_jwt = jwt.encode(to_encode, settings.SECRET_KEY, algorithm=settings.ALGORITHM)
    return encoded_jwt

def decode_access_token(token: str) -> dict:
    try:
        payload = jwt.decode(token, settings.SECRET_KEY, algorithms=[settings.ALGORITHM])
        return payload
    except JWTError:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid authentication token",
            headers={"WWW-Authenticate": "Bearer"},
        )

def get_current_user(credentials: HTTPAuthorizationCredentials = Security(security)) -> dict:
    token = credentials.credentials
    payload = decode_access_token(token)
    user_id = payload.get("sub")
    role = payload.get("role")
    email = payload.get("email")
    if not user_id or not role:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid token payload")
    return {"id": user_id, "role": role, "email": email, "name": payload.get("name", "")}

def require_role(allowed_roles: list[str]):
    def role_checker(current_user: dict = Security(get_current_user)):
        if current_user["role"] not in allowed_roles:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=f"User role '{current_user['role']}' is not authorized. Allowed: {allowed_roles}"
            )
        return current_user
    return role_checker

File: backend/utils/security.js
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const config = require('../config');

// Fast asynchronous password hashing (rounds 8 for sub-10ms speed)
async function hashPasswordAsync(password) {
  return await bcrypt.hash(password, 8);
}

function hashPassword(password) {
  return bcrypt.hashSync(password, 8);
}

// Fast asynchronous password verification (non-blocking)
async function verifyPasswordAsync(plainPassword, hashedPassword) {
  if (!hashedPassword) return false;
  if (plainPassword === hashedPassword) return true;
  try {
    return await bcrypt.compare(plainPassword, hashedPassword);
  } catch (e) {
    return false;
  }
}

function verifyPassword(plainPassword, hashedPassword) {
  if (!hashedPassword) return false;
  if (plainPassword === hashedPassword) return true;
  try {
    return bcrypt.compareSync(plainPassword, hashedPassword);
  } catch (e) {
    return false;
  }
}

function createToken(payload) {
  return jwt.sign(payload, config.SECRET_KEY, { expiresIn: config.JWT_EXPIRES_IN });
}

function authMiddleware(req, res, next) {
  const authHeader = req.headers.authorization;
  if (!authHeader || !authHeader.startsWith('Bearer ')) {
    return res.status(401).json({ detail: 'Invalid or missing authentication token' });
  }

  const token = authHeader.split(' ')[1];
  try {
    const decoded = jwt.verify(token, config.SECRET_KEY);
    req.user = {
      id: decoded.sub,
      role: decoded.role,
      email: decoded.email,
      name: decoded.name
    };
    next();
  } catch (err) {
    return res.status(401).json({ detail: 'Invalid token or session expired' });
  }
}

function requireRole(allowedRoles) {
  return (req, res, next) => {
    authMiddleware(req, res, () => {
      if (!allowedRoles.includes(req.user.role)) {
        return res.status(403).json({
          detail: `User role '${req.user.role}' is not authorized. Allowed: ${allowedRoles.join(', ')}`
        });
      }
      next();
    });
  };
}

module.exports = {
  hashPassword,
  hashPasswordAsync,
  verifyPassword,
  verifyPasswordAsync,
  createToken,
  authMiddleware,
  requireRole
};
